### DISCUSSION TITLE
Unable to install firmware.

### DISCUSSION THREAD

**Discussion by tstr14** [pybricks/support#1953] [General]
I have the LEGO 42146 Liebherr crane and am using a Samsung Galaxy 21 Ultra phone. I have the Pybricks Code set up as an app. Opening the program, I try to install the Py firmware to the 1st hub, following all the steps: Hub is off, Agree to Terms, Name Hub 1, hold hub button until it starts flashing purple, then hit finish. Nothing on the hub changes and no pop-up appears on the phone. If I go to the phone BT settings, I can see LEGO Bootloader, and if I select it, the LED on the hub starts flashing red/green/etc, but a pop up message appears in settings on the phone that I "need an app to use this device". A window does open on the phone with the heading LEGO Bootloader, but it too say app needed to use this device.  Any words of wisdom are GREATLY appreciated.  As Denzel Washington said several times in the movie Philadelphia, explain it to me like I'm a 5 year old, because when it comes to pybricks, I'm at that level or below. Thanks!!!

**Comment by dlech** [comment]
Which web browser are you using?

**Comment by tstr14** [comment]
I'm using Chrome because FF doesn't support the BT function, and I am clicking on the install button - but again, no pop up appears.

**Reply by dlech** [reply]
Are there any error messages in the browser console?

**Comment by tstr14** [comment]
Ah, I didn't look. I only switched to Chrome because of the message that FF wouldn't connect to the BT. I don't normally use Chrome, but I'll see if I can find any error codes.

**Comment by tstr14** [comment]
I can't find any errors messages in chrome.

**Reply by dlech** [reply]
Does the Chrome app have permission to use Bluetooth on your phone?

**Comment by tstr14** [comment]
Yes. I got it to connect and got the popup to pair LEGO Bootloader, but then I got a message telling me the connected hub was different from the one I was trying to load. I KNOW I selected the right hub.......

**Comment by tstr14** [comment]
Ahhhh  it's loading this time! Yay! Now I have to do both hubs, right? Then load the program to both or do I upload that to my controller. Told you I'm a newbie....

**Comment by tstr14** [comment]
Here is the program I am trying to utilize: [https://github.com/orgs/pybricks/discussions/1505](url)

**Reply by laurensvalk** [reply]
To start simple, you might like to try [one of our official examples](https://pybricks.com/project/lego-42146/) before trying the (more advanced) community contributions.

**Reply by tstr14** [reply]
Hi Lauren, thanks for weighing in! I believe I've solved the hub problems, getting the Pybrick firmware to load and then your programs to the upper/lower hubs respectively. Where I'm at now is getting my Xbox controller to work. I cannot get it to connect to my PC, either with a cord or via BT, nor can I get my phone to recognize it via BT, so I'm waiting to borrow another controller to see if it will work. Once I have a controller that I can verify the BT function is operational, I'm assuming all I need to do is pair the upper hub with the controller and I'm good to go - or do I need to pair the lower as well? IIRC, the upper hub issues commands to the lower, so pairing isn't needed?

EDIT: I'm pretty sure I answered all my questions by looking at your instructions again. BTW, here is a picture of my crane:

[img]
![20240317_114300 - Copy-4K](https://github.com/user-attachments/assets/7cd4d346-7493-4e3d-aad4-c2717a53d349)
[/img]

**Comment by tstr14** [comment]
Well, after a week of pulling hair and biting nails, I am about to give up. As near as I can tell, the programs were both successfully installed on the correct hubs but I cannot connect to the hub with the Xbox controller, so nothing works.

**Reply by laurensvalk** [reply]
In cases like these, it can really help to start with something smaller than a very big crane with two hubs.

I can put together an example for you, if you like. Are you using blocks or Python?

Have you meanwhile been able to confirm that your controller has been updated to the latest version using the Xbox or the Xbox accessories app for Windows?

**Reply by BertLindeman** [reply]
> Here is the program I am trying to utilize: [program in discussion 1505](https://github.com/orgs/pybricks/discussions/1505)

Next to what Laurens states here, in the top program I see:
```python

    now_pressed = controller.buttons.pressed()
    if "Y" in now_pressed:
        if light_on:
            light.off()
            light_on = False
            wait(10)
        else:
            light.on(100)
            light_on = True
            wait(10)
        while "Y" in controller.buttons.pressed():  # wait until released
            wait(10)
```

The buttons are no strings (at Version 3.5 and up) but Button values (hope this is the correct term).
In stead of 
```python
    if "Y" in now_pressed:
```

it should be:
```python
    if Button.Y in now_pressed:
```

And maybe other similar changes.

Bert

**Reply by BertLindeman** [reply]
I tried the top and botton programs here and it did **not** work for me at first.

The Technichubs makes it hard to debug.
My mistake was in incorrect motor ports.

After fixing that, the xbox controller connected and the bottom hub light became green.
The joysticks made the motors move.
The switch of the light ON / OFF did not work as I stated in the previous post.

To verify the attachment ports you can move the controller definition 
```python
controller = XboxController()
```
after the definition of the attachments (motors and light).
So potential problems can be seen before the controller makes the hub to disconnect from the PC.
